src/gui_common/CustomDialog.cs

```diff
@@ -0,0 +1,489 @@
+        var titleBarPanel = GetStylebox("custom_titlebar", "WindowDialog");
+        var titleBarHeight = GetConstant("custom_titlebar_height", "WindowDialog");
+
+        // Draw background panels
+        DrawStyleBox(panel, new Rect2(
+            new Vector2(0, -titleBarHeight), new Vector2(RectSize.x, RectSize.y + titleBarHeight)));
+
+        DrawStyleBox(titleBarPanel, new Rect2(
+            new Vector2(3, -titleBarHeight + 3), new Vector2(RectSize.x - 6, titleBarHeight - 3)));
+
+        // Draw title in the title bar
+        var titleFont = GetFont("custom_title_font", "WindowDialog");
+        var titleHeight = GetConstant("custom_title_height", "WindowDialog");
+        var titleColor = GetColor("custom_title_color", "WindowDialog");
+
+        var translated = TranslationServer.Translate(WindowTitle);
+
+        var fontHeight = titleFont.GetHeight() - titleFont.GetDescent() * 2;
+
+        var titlePosition = new Vector2(
+            (RectSize.x - titleFont.GetStringSize(translated).x) / 2, (-titleHeight + fontHeight) / 2);
+
+        DrawString(titleFont, titlePosition, translated, titleColor, (int)(RectSize.x - panel.GetMinimumSize().x));
+
+        // Draw close button highlight
+        if (closeHovered)
+        {
+            var highlight = GetStylebox("custom_close_highlight", "WindowDialog");
+
+            DrawStyleBox(highlight, closeButton.GetRect());
+        }
+    }
+
+    public override void _GuiInput(InputEvent @event)
+    {
+        // Handle title bar dragging
+        if (@event is InputEventMouseButton mouseButton && mouseButton.ButtonIndex == (int)ButtonList.Left)
+        {
+            if (mouseButton.Pressed)
+            {
+                // Begin a possible dragging operation
+                dragType = DragHitTest(new Vector2(mouseButton.Position.x, mouseButton.Position.y));
+
+                if (dragType != DragType.None)
+                    dragOffset = GetGlobalMousePosition() - RectPosition;
+
+                dragOffsetFar = RectPosition + RectSize - GetGlobalMousePosition();
+            }
+            else if (dragType != DragType.None && !mouseButton.Pressed)
+            {
+                // End a dragging operation
+                dragType = DragType.None;
+            }
+        }
+
+        if (@event is InputEventMouseMotion mouseMotion)
+        {
+            if (dragType == DragType.None)
+            {
+                HandlePreviewDrag(mouseMotion);
+            }
+            else
+            {
+                HandleActiveDrag();
+            }
+        }
+    }
+
+    /// <summary>
+    ///   This is overriden so mouse position could take the titlebar into account due to it being drawn
+    ///   outside of the normal Control's rect bounds.
+    /// </summary>
+    public override bool HasPoint(Vector2 point)
+    {
+        var rect = new Rect2(Vector2.Zero, RectSize);
+
+        var titleBarHeight = GetConstant("custom_titlebar_height", "WindowDialog");
```

Comment:
Especially here as I think this HasPoint might get called for each single mouse movement...